File: utilities/scraper.py
import requests
from bs4 import BeautifulSoup
import urllib.parse
from typing import Optional

def is_valid_url(url: str) -> bool:
    """Check if the given string is a valid URL."""
    try:
        result = urllib.parse.urlparse(url)
        return all([result.scheme, result.netloc])
    except ValueError:
        return False

def scrape_url(url: str) -> Optional[str]:
    """
    Fetches the HTML content of a URL and extracts readable text.
    Returns None if the fetch fails.
    """
    if not is_valid_url(url):
        return None

    try:
        # User-Agent headers to avoid basic anti-scraping blocks
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36"
        }
        response = requests.get(url, headers=headers, timeout=10)
        response.raise_for_status()
        
        soup = BeautifulSoup(response.text, 'html.parser')
        
        # Remove noisy elements
        for script_or_style in soup(["script", "style", "header", "footer", "nav", "aside"]):
            script_or_style.decompose()
            
        # Get text
        text = soup.get_text(separator=' ', strip=True)
        # Clean up multiple spaces
        text = ' '.join(text.split())
        return text
        
    except Exception as e:
        print(f"Error scraping {url}: {e}")
        return None

File: lib/scraper.ts
import * as cheerio from 'cheerio';

export async function scrapeUrl(url: string): Promise<string> {
  // If not a URL, treat as raw text
  if (!url.startsWith('http')) return url.slice(0, 5000);

  try {
    const response = await fetch(url, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4103.116 Safari/537.36'
      },
      signal: AbortSignal.timeout(8000), // Vercel Hobby-safe timeout
    });

    if (!response.ok) throw new Error(`HTTP fetch failure: ${response.status}`);

    const html = await response.text();
    const $ = cheerio.load(html);

    // Filter out irrelevant noise
    $('script, style, nav, footer, header, iframe, noscript, svg, path').remove();

    // Prioritize semantic content areas
    const content = $('main, article, [role="main"], #content, .content, .post, .entry').text() || $('body').text();

    // Clean extra whitespace and truncate to keep prompt tokens low
    return content
      .replace(/\s+/g, ' ')
      .trim()
      .slice(0, 5000);
  } catch (error: any) {
    console.error('System Scraper Error:', error.message);
    throw new Error(`Scraper failed: ${error.message}`);
  }
}
